source-mongodb/pull.go

```diff
@@ -79,120 +89,103 @@ func (d *driver) Pull(open *pc.Request_Open, stream *boilerplate.PullOutput) err
-		} else if err == mongo.ErrNoDocuments {
-			// We had previously observed events in the oplog, but now there are none.
-			// This means we have now way of knowing whether we've missed any of the
-			// events, and we'll need to bail out.
-			return fmt.Errorf("the oplog is now empty, but had previously contained entries")
-		} else {
-			log.WithField("oldestOplogEntry", oldestOplogTime).Info("fetched oldest oplog time after completing backfill")
-		}
+	// Persist the checkpoint in full, which may have been updated in updateResourceStates to remove
+	// bindings.
+	if checkpointJson, err := json.Marshal(c.state); err != nil {
+		return fmt.Errorf("marshalling prevState checkpoint: %w", err)
+	} else if err := c.output.Checkpoint(checkpointJson, false); err != nil {
+		return fmt.Errorf("outputting prevState checkpoint: %w", err)
 	}
 
-	// Once we finish the backfill (which may have been completed by a previous instance of the
-	// connector), check to make sure we can start reading the change stream before we started
-	// the backfill. Skip this check if we've already started reading the change stream, or if
-	// we'd only observed an empty change stream before completing the backfill.
-	if len(prevState.StreamResumeToken) == 0 && !prevState.wasChangeStreamEmptyBeforeBackfill() {
-		if _, restartReason := checkBackfillState(&prevState, oldestOplogTime); restartReason != "" {
-			// Automatically restarting the backfill is probably not the best failure mode when
-			// we weren't able to finish the first backfill in time. But until we update this to
-			// work with StateKey (and thus the backfill counter), users won't have an easy way
-			// to do this manually.
-			c.startOverDueToConsistencyLoss(restartReason)
+	didBackfill := false
+	for !prevState.isBackfillComplete(bindings) {
+		didBackfill = true
+		// Repeatedly catch-up reading change streams and backfilling tables for the specified
+		// period of time. This allows resume tokens to be kept reasonably up to date while the
+		// backfill is in progress. The first call to initializeStreams opens change stream cursors,
+		// and streamCatchup catches up the streams to the present and closes the cursors out while
+		// the backfill is in progress.
+		if streams, err := c.initializeStreams(ctx, bindings); err != nil {
+			return err
+		} else if err := c.streamCatchup(ctx, streams); err != nil {
+			return err
+		} else if err := c.backfillCollections(ctx, bindings, backfillFor); err != nil {
+			return err
 		}
 	}
 
-	return c.ChangeStream(ctx, client, bindings, prevState)
-}
-
-func (c *capture) outputCheckpoint(state *captureState, merge bool) error {
-	cp, err := json.Marshal(state)
-	if err != nil {
-		return fmt.Errorf("marshalling checkpoint: %w", err)
-	} else if err = c.Output.Checkpoint(cp, merge); err != nil {
-		return fmt.Errorf("outputting checkpoint: %w", err)
+	if didBackfill {
+		log.Info("backfill complete")
```

Comment:
can we move this log into the block above where we set `didBackfill = true`?

Reply:
It would log on every backfill "round" then. Having it here makes it log "backfill complete" only when the backfills for the collection are complete done. The extra `didBackfill` setting is so that it doesn't log it every time the capture re-starts after completing its backfill.